net/dns/manager.go

```diff
@@ -454,17 +469,87 @@ func (s *dnsTCPSession) handleReads() {
 // servicing DNS requests sent down it.
 func (m *Manager) HandleTCPConn(conn net.Conn, srcAddr netaddr.IPPort) {
 	s := dnsTCPSession{
-		m:            m,
-		conn:         conn,
-		srcAddr:      srcAddr,
-		responses:    make(chan []byte),
-		readClosing:  make(chan struct{}),
+		m:           m,
+		conn:        conn,
+		srcAddr:     srcAddr,
+		responses:   make(chan []byte),
+		readClosing: make(chan struct{}),
 	}
 	s.ctx, s.closeCtx = context.WithCancel(context.Background())
 	go s.handleReads()
 	s.handleWrites()
 }
 
+const dotCertValidity = time.Hour * 24 * 30 // arbitrary; LetsEncrypt-ish
+
+func (m *Manager) dotCert() (tls.Certificate, error) {
+	m.dotCertMu.Lock()
+	defer m.dotCertMu.Unlock()
+
+	if !m.dotCertLast.IsZero() && time.Since(m.dotCertLast) < dotCertValidity {
+		return m.dotCertVal, nil
+	}
+
+	cert, err := genSelfSignedDoTCert()
+	if err == nil {
+		m.dotCertVal = cert
+		m.dotCertLast = time.Now()
+	}
+	return cert, err
+}
+
+// genSelfSignedDoTCert generates a self-signed certificate for DNS-over-TLS
+// (DoT) queries on 100.100.100.100.
+//
+// This exists for Android Private DNS, which in "Automatic" (aka opportunistic)
+// mode doesn't verify certs.
+//
+// See https://github.com/tailscale/tailscale/issues/915.
+func genSelfSignedDoTCert() (tls.Certificate, error) {
+	privKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			Organization: []string{"Tailscale MagicDNS"},
```

Comment:
It might be a good idea to set the common name to something device specific too if such a field is easily obtainable.